cmd/root.go

```diff
@@ -1,133 +1,114 @@
+	var err error
+	if rO.config != "" {
+		cfg, err = config.LoadFromFile(rO.config)
+	} else {
+		cfg, err = config.GetDefault()
+	}
+	if err != nil {
+		return cfg, err
+	}
+	if rO.bind != "" {
+		cfg.Web.Host = rO.bind
+	}
+	if rO.regions != "" {
+		cfg.Regions = strings.Split(rO.regions, ",")
+	}
+	if rO.port != 0 {
+		cfg.Web.Port = rO.port
+	}
+	if rO.prefix != "" {
+		cfg.Web.Prefix = ""
+	}
+	if rO.skipOpen {
+		cfg.Web.SkipOpen = true
+	}
+	if rO.skipRefresh {
+		cfg.Datastore.SkipRefresh = true
+	}
+	return cfg, nil
+}
+
 // NewRootCmd returns the base command when called without any subcommands
 func NewRootCmd(out io.Writer) *cobra.Command {
+	rO := rootOptions{}
 	var rootCmd = &cobra.Command{
 		Use:   "cloudgrep",
 		Short: "A web-based utility to query and manage cloud resources",
 		Long: `Cloudgrep is an app built by RunX to help devops manage the multitude of resources in
 their cloud accounts.`,
 		SilenceUsage: true,
 		RunE: func(cmd *cobra.Command, args []string) error {
-			var cfg config.Config
-			err := viper.Unmarshal(&cfg)
+			cfg, err := rO.loadConfig()
 			if err != nil {
 				return err
 			}
-			err = runCmd(cmd.Context(), cfg, logger)
+			return runCmd(cmd.Context(), cfg, logger)
+		},
+		PersistentPreRun: func(cmd *cobra.Command, args []string) {
+			var err error
+			if rO.verbose {
+				logger, err = zap.NewDevelopment()
+				util.EnableErrorStackTrace()
+			} else {
+				logger, err = zap.NewProduction()
+			}
 			if err != nil {
-				return err
+				panic(err)
 			}
-			return nil
 		},
 	}
 
-	cobra.OnInitialize(initConfig)
-
-	rootCmd.PersistentFlags().StringVarP(&cfgFile, "config", "c", "", "Config file (default is $HOME/.cloudgrep.yaml)")
-	rootCmd.PersistentFlags().BoolVarP(&verbose, "verbose", "v", false, "Log verbosity")
+	rootCmd.PersistentFlags().BoolVarP(&rO.verbose, "verbose", "v", false, "Log verbosity")
 
-	defaultConfig, _ := config.GetDefault()
 	flags := rootCmd.Flags()
 
-	flags.String("bind", defaultConfig.Web.Host, "Host to bind on")
-	_ = viper.BindPFlag("web.host", flags.Lookup("bind"))
-
-	flags.StringP("regions", "r", "", "Comma separated list of regions to scan, or \"all\"")
-	_ = viper.BindPFlag("regions", flags.Lookup("regions"))
-
-	flags.IntP("port", "p", defaultConfig.Web.Port, "Port to use")
-	_ = viper.BindPFlag("web.port", flags.Lookup("port"))
-
-	flags.String("prefix", defaultConfig.Web.Prefix, "URL prefix to use")
-	_ = viper.BindPFlag("web.prefix", flags.Lookup("prefix"))
-
-	flags.Bool("skip-open", defaultConfig.Web.SkipOpen, "Skip running the open command to open default browser")
-	_ = viper.BindPFlag("web.skipOpen", flags.Lookup("skip-open"))
-
-	flags.Bool("skip-refresh", defaultConfig.Datastore.SkipRefresh, "Skip running data refresh on start up")
-	_ = viper.BindPFlag("datastore.skipRefresh", flags.Lookup("skip-refresh"))
-
-	rootCmd.AddCommand(NewVersionCommand(out))
+	flags.StringVarP(&rO.config, "config", "c", "", "Config file (default is $HOME/.cloudgrep.yaml)")
```

Comment:
in which scenario the default file `$HOME/.cloudgrep.yaml` is used?
The way I understand it:
- default is https://github.com/run-x/cloudgrep/blob/main/pkg/config/config.yaml
- if the user sets `--config` then a value is required so we won't use this default path